src/main/java/fr/sncf/osrd/infra/trackgraph/Switch.java

```diff
@@ -1,26 +1,83 @@
 package fr.sncf.osrd.infra.trackgraph;
 
 import fr.sncf.osrd.infra.signaling.Signal;
+import fr.sncf.osrd.utils.graph.EdgeEndpoint;
 
+import java.util.List;
+import java.util.Map;
 import java.util.ArrayList;
 
 public class Switch extends TrackNode {
     public final int switchIndex;
-    public final double positionChangeDelay;
-    public TrackSection leftTrackSection;
-    public TrackSection rightTrackSection;
+    public final double groupChangeDelay;
+    public final List<Port> ports;
+    public final Map<String, List<PortEdge>> groups;
     public ArrayList<Signal> signalSubscribers;
 
     Switch(
             TrackGraph graph,
             int index,
             String id,
             int switchIndex,
-            double positionChangeDelay) {
+            double groupChangeDelay,
+            List<Port> ports,
+            Map<String, List<PortEdge>> groups) {
         super(index, id);
         this.switchIndex = switchIndex;
-        this.positionChangeDelay = positionChangeDelay;
+        this.groupChangeDelay = groupChangeDelay;
+        this.ports = ports;
+        this.groups = groups;
         this.signalSubscribers = new ArrayList<>();
         graph.registerNode(this);
     }
+
+    public String getDefaultGroup() {
+        return groups.entrySet().stream().findFirst().get().getKey();
+    }
+
+    public static final class Port {
+
+        public final String id;
+        public final TrackSection trackSection;
+        public final EdgeEndpoint endpoint;
+
+        /**
+         * Create a new switch port 
+         * @param id the id of the port
+         * @param trackSection the track section connected to the port
+         * @param endpoint the endpoint of the connected track section
+         */
+        public Port(String id, TrackSection trackSection, EdgeEndpoint endpoint) {
+            this.id = id;
+            this.trackSection = trackSection;
+            this.endpoint = endpoint;
+        }
+
+        @Override
+        public int hashCode() {
+            return id.hashCode();
+        }
+
+        @Override
+        public boolean equals(Object p) {
+            if (p == null)
+                return false;
+            if (p.getClass() != Port.class)
+                return false;
+            if (id == null)
+                throw new NullPointerException();
```

Comment:
Not checking anything here would also result in a NullPointerException anyway. If we don't allow null ids, it should be checked in the constructor.